### DISCUSSION TITLE
Pass variables as "secret" payload

### DISCUSSION THREAD

**Discussion by freebuu** [slack-php/slack-php-app-framework#8] [Q&A] [answered]
Hello everyone!
I have this case:

1. User fill fields in modal 
2. On submit - I update editing modal to "preview" modal - where user can check their information before submit
3. On submit preview modal - I want to work with information

But on step 3 I have just text - which I show to user. But this is not for work - just for show.

How can I pass variables on step 2 to modal on step 3?
something like
```
//on viewSubmission edit modal
$previewModal = Modal::new()
//other vars
  ->tap(function (Modal $modal){
      $modal
            ->addPayloadVar('var1', 'value1')
            ->addPayloadVar('var2', 'value2')
  });
$ctx->view()->update($previewModal);

//on viewSubmission preview modal
$ctx->payload()->get('var1');
```

**Comment by jlindblom-godaddy** [comment, marked as answer]
This is a good use case for the modal's `private_metadata` field (see https://api.slack.com/reference/surfaces/views).

Within the slack-php framework, there is a simple abstraction for this that is very much like what you've requested.

To set (during step 2):
```php
$previewModal = Modal::new()
    // ...
    ->tap(function (Modal $modal){
        $modal->encodePrivateMetadata([
            'var1' => 'value1',
            'var2' => 'value2',
        ]);
        // ...
    });
```

To get (during step 3):
```php
$ctx->payload()->getMetadata()->get('var1');
```

**Note**: If you inspect the `private_metadata` in the modal's JSON, you will see a base64-encoded string. The framework does `http_build_query` and `base64_encode` to the data you provide to `encodePrivateData`. `getMetadata()` does the reverse of that. This let's you send funky data (e.g., json-encoded strings) safely through the `private_metadata` fields.

---

**Another Note**: There is a way to use `private_metadata` without the abstraction too, if you just want to send a raw string (which I've personally found less useful, which is why I added the abstraction in the first place):
- Set: `$modal->privateMetadata($rawData)`
- Get: `$ctx->payload()->get('view.private_metadata')`

**Comment by freebuu** [comment]
@jlindblom-godaddy wow, that's it, thank you! I looking in documentation whole hours, but did't find this elem :(
